// @midnight-ntwrk/midnight-js-types/dist/index.mjs
export { UnprovenTransaction } from '@midnight-ntwrk/ledger';

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Typesafe version of `Object.keys(contract.impureCircuits)`.
 *
 * @param contract The contract having impure circuits for which we want ids.
 *
 * @typeParam C The contract type for which we would like impure circuit IDs.
 */
const getImpureCircuitIds = (contract) => Object.keys(contract.impureCircuits);

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * An error describing an invalid protocol scheme.
 */
class InvalidProtocolSchemeError extends Error {
    invalidScheme;
    allowableSchemes;
    /**
     * @param invalidScheme The invalid scheme.
     * @param allowableSchemes The valid schemes that are allowed.
     */
    constructor(invalidScheme, allowableSchemes) {
        super(`Invalid protocol scheme: '${invalidScheme}'. Allowable schemes are one of: ${allowableSchemes.join(',')}`);
        this.invalidScheme = invalidScheme;
        this.allowableSchemes = allowableSchemes;
    }
}

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * A valid named log level.
 */
var LogLevel;
(function (LogLevel) {
    /**
     * Log levels typically used by DAapp developers.
     */
    LogLevel["INFO"] = "info";
    LogLevel["WARN"] = "warn";
    LogLevel["ERROR"] = "error";
    LogLevel["FATAL"] = "fatal";
    /**
     * Log levels used by Midnight.JS to report internal state.
     */
    LogLevel["DEBUG"] = "debug";
    LogLevel["TRACE"] = "trace";
})(LogLevel || (LogLevel = {}));

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Creates a branded prover key representation from a prover key binary.
 *
 * @param uint8Array The prover key binary.
 */
const createProverKey = (uint8Array) => {
    return uint8Array;
};
/**
 * Creates a branded verifier key representation from a verifier key binary.
 *
 * @param uint8Array The verifier key binary.
 */
const createVerifierKey = (uint8Array) => {
    return uint8Array;
};
/**
 * Creates a branded ZKIR representation from a ZKIR binary.
 *
 * @param uint8Array The ZKIR binary.
 */
const createZKIR = (uint8Array) => {
    return uint8Array;
};
/**
 * Creates an {@link UnbalancedTransaction} from a ledger transaction.
 *
 * @param tx The ledger transaction to wrap.
 */
const createUnbalancedTx = (tx) => {
    return tx;
};
/**
 * Creates an {@link BalancedTransaction} from a ledger transaction.
 * @param tx The ledger transaction to wrap.
 */
const createBalancedTx = (tx) => {
    return tx;
};
/**
 * Indicates that the transaction is invalid.
 */
const FailEntirely = 'FailEntirely';
/**
 * Indicates that the transaction is valid but the portion of the transcript
 * that is allowed to fail (the portion after a checkpoint) did fail. All effects
 * from the guaranteed part of the transaction are kept but the effects from the
 * fallible part of the transaction are discarded.
 */
const FailFallible = 'FailFallible';
/**
 * Indicates that the guaranteed and fallible portions of the transaction were
 * successful.
 */
const SucceedEntirely = 'SucceedEntirely';

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * A provider for zero-knowledge intermediate representations, prover keys, and verifier keys. All
 * three are used by the {@link ProofProvider} to create a proof for a call transaction. The implementation
 * of this provider depends on the runtime environment, since each environment has different conventions
 * for accessing static artifacts.
 * @typeParam K - The type of the circuit ID used by the provider.
 */
class ZKConfigProvider {
    /**
     * Retrieves the verifier keys produced by `compactc` for the given circuits.
     * @param circuitIds The circuit IDs of the verifier keys to retrieve.
     */
    async getVerifierKeys(circuitIds) {
        return Promise.all(circuitIds.map(async (id) => {
            const key = await this.getVerifierKey(id);
            return [id, key];
        }));
    }
    /**
     * Retrieves all zero-knowledge artifacts produced by `compactc` for the given circuit.
     * @param circuitId The circuit ID of the artifacts to retrieve.
     */
    async get(circuitId) {
        return {
            circuitId,
            proverKey: await this.getProverKey(circuitId),
            verifierKey: await this.getVerifierKey(circuitId),
            zkir: await this.getZKIR(circuitId)
        };
    }
}

export { FailEntirely, FailFallible, InvalidProtocolSchemeError, LogLevel, SucceedEntirely, ZKConfigProvider, createBalancedTx, createProverKey, createUnbalancedTx, createVerifierKey, createZKIR, getImpureCircuitIds };
